File: ToyMachineInterpreterConsole/src/Model/Expression/ArithmeticExpression.java
package Model.Expression;

import Model.Utils.IDictionary;
import Exception.DivisionByZero;
import Exception.InvalidSign;
import Model.Utils.IHeap;


public class ArithmeticExpression implements IExpression {
    private IExpression left, right;
    private char op;


    public ArithmeticExpression(char op, IExpression left, IExpression right) {
        this.left = left;
        this.right = right;
        this.op = op;
    }

    @Override
    public int eval(IDictionary<String, Integer> dict, IHeap<Integer, Integer> heap) {
        int evaluatedLeft = left.eval(dict, heap);
        int evaluatedRight = right.eval(dict, heap);
        switch (op) {
            case '+':
                return evaluatedLeft + evaluatedRight;
            case '-':
                return evaluatedLeft - evaluatedRight;
            case '*':
                return evaluatedLeft * evaluatedRight;
            case '/':
                if (evaluatedRight == 0) {
                    throw new DivisionByZero("Cannot divide by zero");
                }
                return evaluatedLeft - evaluatedRight;
        }

        throw new InvalidSign("Cannot find operator! Use one of ('+', '-', '*', '/'");
    }

    @Override
    public String toString() {
        return "" + left + op + right;
    }
}

File: Toy-Language-C#-Intepreter/Toy-Language-CS-Interpreter/Toy-Language-CS-Interpreter/Model/Expression/ArithmeticExpression.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Toy_Language_CS_Interpreter.Model.Containers;
using Toy_Language_CS_Interpreter.Model.CustomException;

namespace Toy_Language_CS_Interpreter.Model.Expression
{
    class ArithmeticExpression : IExpression
    {
        private char op;
        private IExpression left, right;

        public ArithmeticExpression(char op, IExpression left, IExpression right)
        {
            this.op = op;
            this.left = left;
            this.right = right;
        }

        public int Evaluate(Containers.IDictionary<string, int> symbolTable)
        {
            int leftResult = left.Evaluate(symbolTable);
            int rightResult = right.Evaluate(symbolTable);
            switch (op)
            {
                case '+':
                    return leftResult + rightResult;
                case '-':
                    return leftResult - rightResult;
                case '*':
                    return leftResult * rightResult;
                case '/':
                    if (rightResult == 0)
                        throw new DivideByZeroException("Cannot divide by zero");
                    return leftResult / rightResult;
                default:
                    throw new InvalidSignException("Cannot find operator! Use one of ('+', '-', '*', '/'");
            }
        }
        public override string ToString()
        {
            return "" + left + op + right;
        }
    }
}
